### PR TITLE
Handle adding and removing deps in AssetsDefinition.map_asset_specs

### PR DESCRIPTION
**Pull request by dpeng817**
# Background
This PR changes `AssetsDefinition.map_asset_specs` to allow adding additional deps to an `AssetSpec`.

To do this, we create `Nothing` inputs under the hood which map to the asset dependency in question.

This is important mostly for integration cases where we use the decorator pattern `@dbt_assets`, `@fivetran_assets`, etc. If you want to link an asset dependency to assets defined with those decorators, it's currently not feasible for users.

# Caveats

**Only supports op-backed assets**
This PR adds support only for op-backed assets, not graph-backed assets. I attempted to add support and found myself blocked since graph-backed assets currently do not support non-argument deps (at least from what I could gather). I added a linear issue to track this and referenced it in the relevant place in code.

**Doesn't work for argument-based deps**
This PR adds support for the deletion of non-argument based deps; but does not support the deletion of argument-based deps. I think deletion of argument-based deps is impossible, since you would have nothing to pass to the underlying function. We error cleanly instead, and I think that's the best we can do.

# Testing

A bunch of tests for every case that I could think of:
- add additional dependencies
- add the same dependency multiple times
- remove a dependency
- add dependencies with partition mappings
- add dependencies to a non-executable assets def
- attempt to add deps to a graph-backed asset

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26144** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26144?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26141** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 1]
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+
+
+def _thread_deps_to_node(
+    assets_def: AssetsDefinition, additional_deps: Mapping[AssetKey, AbstractSet[AssetDep]]
+) -> "AssetsDefinition":
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    if not assets_def.is_executable:
+        return assets_def
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1918]: why the early return?

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 2]
```diff
@@ -1286,27 +1290,32 @@ def update_replace_dict_and_conflicts(
...
-                    "must not change deps."
-                )
+
+            deps_per_key = defaultdict(list)
+            for dep in mapped_spec.deps:
+                deps_per_key[dep.asset_key].append(dep)
+            for dep in spec.deps:
+                if dep.asset_key not in deps_per_key:
+                    raise DagsterInvalidDefinitionError(
+                        f"After mapping, asset {dep.asset_key.to_user_string()} was not found in the mapped dependencies. You should not remove dependencies from assets, since that can destroy the internal structure of the asset."
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1308]: hmmmm why not let the user do this? I'm not really following why this should always be illegal

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 3]
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+    assets_def: AssetsDefinition, additional_deps: Mapping[AssetKey, AbstractSet[AssetDep]]
+) -> "AssetsDefinition":
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    if not assets_def.is_executable:
+        return assets_def
+    if not isinstance(assets_def.node_def, OpDefinition):
+        raise DagsterInvalidDefinitionError("Cannot add dependencies to a graph-backed asset")
+    additional_input_names_by_key = {
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1921]: super confusing variable name since it is actually indexed by name

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 4]
```diff
@@ -85,6 +85,10 @@
 ASSET_SUBSET_INPUT_PREFIX = "__subset_input__"
 
 
+def stringify_asset_key_to_input_name(asset_key: AssetKey) -> str:
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:88]: is there not an existing utility that does this? We should use a single, canonical one

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 5]
```diff
@@ -1897,3 +1906,70 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+        for dep in deps
+    }
+
+    complete_input_names_by_key = {
+        **assets_def.node_keys_by_input_name,
+        **additional_input_names_by_key,
+    }
+
+    additional_ins = {
+        input_name: In(dagster_type=Nothing) for input_name in additional_input_names_by_key
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1933]: given the name/key inversion this is basically unreviewable so will just comment on the idea

**Review by schrockn — DISMISSED** [review]
It makes sense to support this.

One point of discussion is the behavior of `merge_attributes` w/r/t to deps. My initial thought is to make it work like the `tags` and `metadata` where collission are just overwritten. Essentially treat it like a dictionary indexed by asset key.

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 1 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1918]: there's no op if the asset isn't executable, and attempting to access the property throws an error - so we shouldn't be attempting any op-level munging.

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1921]: oof. Yea I'll fix that

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 2 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1308]: Hmmm that does not match up to the description in the error message. If this is for  arguemnt-based deps that should be explained in the errror message. However this should work for vanilla deps.

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 6]
```diff
@@ -1897,3 +1910,74 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+    """
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    check.invariant(
+        assets_def.is_executable, "Cannot thread deps to node for non-executable asset."
+    )
+    if not isinstance(assets_def.node_def, OpDefinition):
+        raise DagsterInvalidDefinitionError("Cannot add dependencies to a graph-backed asset")
+    dep_key_by_input_name = {
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1929]: nit: generally like blank lines after dedents in Python for readability

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 7]
```diff
@@ -1897,3 +1910,74 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
     """
     sorted_key_strs = sorted(str(key) for key in entity_keys)
     return non_secure_md5_hash_str(json.dumps(sorted_key_strs).encode("utf-8"))[:8]
+
+
+def _thread_deps_to_node(
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1915]: function name a bit add.

`_assets_def_with_additional_deps` or something like that more clear

**Review by schrockn — CHANGES_REQUESTED** [review]
Ok this is looking better.

A few things

* Remove the RFC tag from the PR title if this is meant for final review
* Rewrite the summary to be worthy of a real commit. Correct, proper English, etc.
* We can't just not do `graph_asset` because " I don't think support would be that hard to add, just involves a bunch of additional gross munging that I don't really want to do, and graph-backed assets have limited adoption anyway."

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 2 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1308]: I'll have to dig into whether there's a way to disambiguate argument based deps from non-argument based deps.

**Comment by dpeng817** [comment]
Okay, now that I've validated direction and incorporated feedback, I think this PR is ready for final review. I think I've addressed all outstanding comments.

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 2 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py]: I was able to find a way to disambiguate, so I think we're good to support this 👍

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 7 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py]: changed the name, also should reflect the fact that we're potentially deleting deps now too

**Comment by dpeng817** [comment]
@schrockn ; given your current increased responsibilities and given that I've addressed all outstanding comments, I'll have @OwenKephart review this one to the finish line.

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 8]
```diff
@@ -1897,3 +1896,108 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+    """
+    from dagster._builtins import Nothing
+    from dagster._core.definitions.input import In
+
+    # If the assets_def is not executable (and therefore is not backed by a node definition), return the original assets_def
+    if not assets_def.is_executable:
+        return assets_def
+    original_key_to_input_mapping = reverse_dict(assets_def.node_keys_by_input_name)
+    # Get the new deps that are not already in the original mapping
+    new_deps = {
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1915]: This code is a bit hard to follow, which is partly unavoidable as we have all these annoying dictionaries, but the way I'm thinking about it, it might be easier to handle this purely as sets to start with, i.e.:

```python
# mapped_specs is the list of specs that were mapped in the above fn
new_deps = set().union(*(spec.deps for spec in mapped_specs))
previous_deps = set().union(*(spec.deps for spec in original_assets_def.specs))

added_deps = new_deps-previous_deps
deleted_deps = previous_deps-new_deps

if not added_deps and not deleted_deps: ...
```

Possible that I'm missing some subtle or not-so-subtle stuff that makes that pattern not work, so if that's the case feel free to ignore

**Review by OwenKephart — CHANGES_REQUESTED** [review]
agree this should be added, have some comments on code structure here

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 9]
```diff
@@ -1897,3 +1896,108 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+    ins_from_input_defs = {
+        input_def.name: In.from_definition(input_def) for input_def in all_input_defs
+    }
+    new_node_def = assets_def.op.with_replaced_properties(
+        name=assets_def.op.name,
+        ins=ins_from_input_defs,
+    )
+
+    updated_dep_keys = {key: {dep.asset_key for dep in deps} for key, deps in updated_deps.items()}
+    return AssetsDefinition(
```
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1982]: AssetsDefinition.dagster_internal_init explicitly doesn't allow you to pass both attributes dict _and_ specs, which is why we need to do the double-assetsdefinition-construction thing here.

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 10]
```diff
@@ -1286,27 +1290,22 @@ def update_replace_dict_and_conflicts(
...
-                raise DagsterInvalidDefinitionError(
-                    f"Asset deps {spec.deps} were changed to {mapped_spec.deps}. Mapping function "
-                    "must not change deps."
-                )
 
+            all_deps_per_asset_key[spec.key] = set(mapped_spec.deps)
             mapped_specs.append(mapped_spec)
 
-        return self.__class__.dagster_internal_init(
+        assets_def_new_specs = self.__class__.dagster_internal_init(
```
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1305]: I can take a pass at restructuring. The reason I did this way was to avoid re-building the asset dependency structure, but perhaps that's actually unavoidable anyway. I do so much dictionary munging anyway I'm not convinced I've actually saved any time complexity

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1982]: not quite following this -- the existing code was already calling `self.__class__.dagster_internal_init(**{**self.get_attributes_dict(), "specs": mapped_specs})`, and the only difference here seems to be that we're changing `keys_by_input_name`, which is a valid parameter for `dagster_internal_init`. So what goes wrong when you do:

```
self.__class__.dagster_internal_init(
    **{**self.get_attributes_dict(), "specs": mapped_specs, "keys_by_input_name": ...}
)
```
?

[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1982]: anyway the concern here (beyond it being verbose) is that this is sorta a ticking timebomb if we ever add an additional parameter to AssetsDefinition and forget to add it here

[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1982]: Ah - turns out I was wrong. I was under the impression that you could _not_ set both input-to-asset-key mappings and specs, but it turns out you can.

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 8 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1915]: quality suggestion. I think the core complication comes from me trying to save a linear pass over the replaced specs. Which is .... dum

File: python_modules/dagster/dagster/_core/definitions/assets.py [hunk 11]
```diff
@@ -1897,3 +1895,64 @@ def unique_id_from_asset_and_check_keys(entity_keys: Iterable["EntityKey"]) -> s
...
+    from dagster._core.definitions.input import In
+
+    new_deps = set().union(*(spec.deps for spec in replaced_specs))
+    previous_deps = set().union(*(spec.deps for spec in assets_def.specs))
+    added_deps = new_deps - previous_deps
+    removed_deps = previous_deps - new_deps
+    remaining_original_deps = previous_deps - removed_deps
+    original_key_to_input_mapping = reverse_dict(assets_def.node_keys_by_input_name)
+
+    # If there are no changes to the dependency structure, return the original assets_def
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/_core/definitions/assets.py:1913]: nit: slightly confusing comment as we're not returning the original assets_def (we're returning an assets def with the original node)

**Review by OwenKephart — APPROVED** [review]
excellent

File: python_modules/dagster/dagster/_core/definitions/assets.py (hunk 11 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py:1913]: oof you right

Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/assets.py]: fixed
